File: src/keystore_interaction.py
import os
import time
import datetime
import pickle
import logging

KEYSTORE_PREFIX = ''
KEYSTORE_SUFFIX = '.keystore'
DEFAULT_PATH = './'


class KeystoreBox:
    def __init__(self, Path_, KeystoreData_):
        self.Path = Path_
        self.KeystoreData = KeystoreData_

class KeystoreMaster:
    keystoreBox = None

    def __init__(self, KeystorePath_=None):
        if KeystorePath_ is not None:
            k = self.tryParseKeystoreFile(KeystorePath_)
            if k is not None:
                self.keystoreBox = KeystoreBox(KeystorePath_, k)

        if self.keystoreBox is None:
            kbox = self.searchForKeystore()
            if kbox is None:
                self.keystoreBox = self.createKeystore()
            else:
                self.keystoreBox = kbox

    def searchForKeystore(self, search_dir=DEFAULT_PATH):
        try:
            for filename in os.listdir(search_dir):
                if filename.endswith(KEYSTORE_SUFFIX):
                    k = self.tryParseKeystoreFile(search_dir+filename)
                    if k is not None:
                        kbox = KeystoreBox(search_dir+filename, k)
                        logging.debug('Keystore successfully found %s', kbox.Path)
                        return kbox
            return None
        except:
            logging.debug('Keystore search exception')
            return None

    def createKeystore(self, keystore_path=None):
        if keystore_path is None:
            keystore_path = DEFAULT_PATH + self.createKeystoreName()
        else:
            keystore_path += KEYSTORE_SUFFIX

        keystore_data = {
            'created': self.timestamp(),
            'uploads_info': {}
        }

        if self.marshal(keystore_path, keystore_data):
            logging.debug('Keystore created successfully %s', keystore_path)
            return KeystoreBox(keystore_path, keystore_data)
        else:
            logging.debug('Keystore creation exception')
            return None

    def tryParseKeystoreFile(self, filePath):
            data = self.unmarshal(filePath)

            if data is None or data.get('created') is None:
                logging.debug('Validation is not successful')
                return None
            logging.debug('Validation is successful')
            return data

    def getListOfUploadedFiles(self):
        print(list(self.keystoreBox.KeystoreData['uploads_info'].keys()))
        return list(self.keystoreBox.KeystoreData['uploads_info'].keys())

    def getUploadPiecesInfo(self, fileName):
        return self.keystoreBox.KeystoreData['uploads_info'].get(fileName)

    def createKeystoreName(self):
        return KEYSTORE_PREFIX + self.timestamp() + KEYSTORE_SUFFIX

    def isFilenameCollide(self, filename):
        if self.keystoreBox.KeystoreData['uploads_info'].get(filename) is None:
            return False
        else:
            return True

    def isFileCollide(self, filePath):
        fileName = os.path.basename(filePath)
        return self.isFilenameCollide(fileName)

    def addUploadedFileToKeystore(self, filename, upload_info):
        if self.keystoreBox is not None:
            if not self.isFilenameCollide(filename):
                self.keystoreBox.KeystoreData['uploads_info'][filename] = upload_info
                return self.marshal(self.keystoreBox.Path, self.keystoreBox.KeystoreData)
            else:
                logging.debug('Upload with same filename %s', filename)
                return False
        else:
            logging.debug('Keystore file is not present')
            return False

    def catKeystoreFile(self):
        print(self.tryParseKeystoreFile(self.keystoreBox.Path))

    def marshal(self, filePath, data):
        try:
            file = open(filePath, "wb")
            pickle.dump(data, file)
            file.close()
            return True
        except:
            logging.debug('Marshal exception')
            return False

    def unmarshal(self, filePath):
        try:
            file = open(filePath, "rb")
            data = pickle.load(file)
            file.close()
            return data
        except:
            logging.debug('Unmarshal exception')
            return None

    def timestamp(self):
            ts = time.time()
            return datetime.datetime.fromtimestamp(ts).strftime('%Y-%m-%d %H:%M:%S')

File: src/keystore_interaction.ts
import {IUploadedFile} from "./blockchain_interaction";
import {readdirSync, readFileSync, writeFileSync} from "fs";
import {sep} from "path";

export interface IKeystoreFile {
    filePath: string;
    data: IKeystoreData;
}
export interface IKeystoreData {
    version: number;
    created: number;
    uploadedFiles: IUploadedFile[];
}

const KEYSTORE_PREFIX = '';
const KEYSTORE_SUFFIX = '.keystore';
const DEFAULT_PATH = '.'+sep;
const CURRENT_KEYSTORE_VERSION = 0;

export class KeystoreMaster {
    keystoreFile: IKeystoreFile;

    constructor(searchPath:string=DEFAULT_PATH) {
        let keystoreFilePath = this.searchForKeystore(searchPath);
        if(keystoreFilePath){
            let k: Buffer = readFileSync(keystoreFilePath);
            this.keystoreFile = {
                filePath: keystoreFilePath,
                data: this.unmarshal(k),
            };
        } else {
            this.keystoreFile = this.createKeystore(DEFAULT_PATH);
            this.saveKeystoreFile();
        }
    }

    searchForKeystore(searchPath:string):string | null {
        let files: string[] = readdirSync(searchPath);
        for(let file of files) {
            if(file.endsWith(KEYSTORE_SUFFIX)) return searchPath+file
        }
        return null;
    }
    createKeystore(creationFolder:string):IKeystoreFile{
        let keystoreData: IKeystoreData = {
            version: CURRENT_KEYSTORE_VERSION,
            created: new Date().getTime(),
            uploadedFiles: [],
        };
        let fileName: string = this.createKeystoreName();
        return {
            filePath: creationFolder+fileName,
            data: keystoreData,
        }
    }
    createKeystoreName():string {
        return KEYSTORE_PREFIX+this.getTimestamp()+KEYSTORE_SUFFIX;
    }
    saveKeystoreFile() {
        let k: Buffer = this.marshal(this.keystoreFile.data);
        writeFileSync(this.keystoreFile.filePath, k);
    }
    getUploadHistory():string[]{
        return this.keystoreFile.data.uploadedFiles.map(a => a.fileName);
    }
    getUploadedFileInfo(filename:string):IUploadedFile | null {
        for(let file of this.keystoreFile.data.uploadedFiles){
            if(file.fileName === filename)
                return file
        }
        return null
    }
    addUploadedFileToKeystore(file: IUploadedFile){
        this.keystoreFile.data.uploadedFiles.push(file);
        this.saveKeystoreFile();
    }
    marshal(keystoreData: IKeystoreData):Buffer {
        let k: string = JSON.stringify(keystoreData);
        return Buffer.from(k, "UTF-8")
    }
    unmarshal(keytoreData: Buffer):IKeystoreData{
        let k: string = keytoreData.toString("UTF-8");
        return JSON.parse(k);
    }
    //todo сделать util класс для этого
    getTimestamp():string {
        let currentDate = new Date();

        let h = currentDate.getHours();
        let m = currentDate.getMinutes();
        let s = currentDate.getSeconds();
        let d = currentDate.getDate();
        let M = currentDate.getMonth() + 1;
        let y = currentDate.getFullYear();

        return d + "-" + M + "-" + y + " " + h + ":" + m + ":" + s;
    }
}